File: tools/openrouter-zig/src/budget.zig
const std = @import("std");
const models = @import("models.zig");
const DEFAULT_BUDGET: f64 = 6.00;
extern "c" fn getenv([*:0]const u8) ?[*:0]const u8;
fn budgetPath(allocator: std.mem.Allocator) ![]const u8 {
    const home = if (getenv("HOME")) |h| std.mem.span(h) else ".";
    return std.fs.path.join(allocator, &.{ home, ".orcli_budget" });
}
pub fn readBudget(io: std.Io, allocator: std.mem.Allocator) !models.BudgetState {
    const path = try budgetPath(allocator);
    defer allocator.free(path);
    const content = std.Io.Dir.cwd().readFileAlloc(io, path, allocator, @enumFromInt(1024)) catch {
        return models.BudgetState{ .remaining = DEFAULT_BUDGET, .cap = DEFAULT_BUDGET };
    };
    defer allocator.free(content);
    const value = std.fmt.parseFloat(f64, std.mem.trim(u8, content, " \n\r")) catch DEFAULT_BUDGET;
    return models.BudgetState{ .remaining = value, .cap = DEFAULT_BUDGET };
}
pub fn writeBudget(io: std.Io, allocator: std.mem.Allocator, remaining: f64) !void {
    const path = try budgetPath(allocator);
    defer allocator.free(path);
    const buf = try std.fmt.allocPrint(allocator, "{d:.4}", .{remaining});
    defer allocator.free(buf);
    const file = try std.Io.Dir.createFileAbsolute(io, path, .{});
    defer file.close(io);
    try file.writePositionalAll(io, buf, 0);
}
pub fn trackCost(
    io: std.Io,
    allocator: std.mem.Allocator,
    prompt_tokens: u32,
    completion_tokens: u32,
    prompt_cost_per_m: f64,
    completion_cost_per_m: f64,
) !models.BudgetState {
    const cost = (@as(f64, @floatFromInt(prompt_tokens)) * prompt_cost_per_m +
        @as(f64, @floatFromInt(completion_tokens)) * completion_cost_per_m) / 1_000_000.0;
    const current = try readBudget(io, allocator);
    const remaining = current.remaining - cost;
    try writeBudget(io, allocator, remaining);
    return models.BudgetState{ .remaining = remaining, .cap = current.cap };
}
pub fn formatBudget(allocator: std.mem.Allocator, state: models.BudgetState) ![]const u8 {
    return std.fmt.allocPrint(allocator, "${d:.4} remaining · cap ${d:.2}", .{ state.remaining, state.cap });
}

File: tools/openrouter-ts/src/budget.ts
"use strict";
import { homedir } from "node:os";
import { join } from "node:path";
import { readFileSync, writeFileSync, existsSync } from "node:fs";
import type { BudgetState } from "./types.js";
const BUDGET_FILE = join(homedir(), ".orcli_budget");
const DEFAULT_BUDGET = 6.0;
export function readBudget(): BudgetState {
  try {
    if (existsSync(BUDGET_FILE)) {
      const remaining = parseFloat(readFileSync(BUDGET_FILE, "utf-8").trim());
      return {
        remaining,
        spent: DEFAULT_BUDGET - remaining,
        cap: DEFAULT_BUDGET,
      };
    }
  } catch {
  }
  return { remaining: DEFAULT_BUDGET, spent: 0, cap: DEFAULT_BUDGET };
}
export function writeBudget(remaining: number): void {
  writeFileSync(BUDGET_FILE, remaining.toFixed(4));
}
export function trackCost(
  promptTokens: number,
  completionTokens: number,
  promptCostPerM: number,
  completionCostPerM: number,
): BudgetState {
  const cost =
    (promptTokens * promptCostPerM) / 1_000_000 +
    (completionTokens * completionCostPerM) / 1_000_000;
  const current = readBudget();
  const remaining = current.remaining - cost;
  writeBudget(remaining);
  return {
    remaining,
    spent: current.spent + cost,
    cap: current.cap,
  };
}
export function formatBudget(state: BudgetState): string {
  return `$${state.remaining.toFixed(4)} remaining · $${state.spent.toFixed(4)} spent · cap $${state.cap.toFixed(2)}`;
}
export function affordableTokens(
  promptCostPerM: number,
  completionCostPerM: number,
): number {
  const { remaining } = readBudget();
  if (remaining <= 0) return 0;
  return Math.floor((remaining / completionCostPerM) * 1_000_000);
}